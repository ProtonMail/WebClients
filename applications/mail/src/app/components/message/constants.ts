export const MESSAGE_IFRAME_ROOT_ID = 'proton-root';
export const MESSAGE_IFRAME_BLOCKQUOTE_ID = 'proton-blockquote';
export const MESSAGE_IFRAME_TOGGLE_ID = 'proton-toggle';
export const MESSAGE_IFRAME_PRINT_ID = 'proton-print';
