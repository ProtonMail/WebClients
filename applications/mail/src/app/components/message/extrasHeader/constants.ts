export const MESSAGE_LABEL_DROPDOWN_ID = 'message-header-expanded:label-dropdown';
export const MESSAGE_FOLDER_DROPDOWN_ID = 'message-header-expanded:folder-dropdown';
export const MESSAGE_FILTER_DROPDOWN_ID = 'message-header-expanded:filter-dropdown';
