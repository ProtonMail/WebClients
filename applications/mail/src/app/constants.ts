import { MAILBOX_LABEL_IDS } from '@proton/shared/lib/constants';
import { c } from 'ttag';
import { ESCache, ESIndexingState, ESStatus } from './models/encryptedSearch';

export const MAIN_ROUTE_PATH = '/:labelID?/:elementID?/:messageID?';

export const EXPIRATION_CHECK_FREQUENCY = 10000; // each 10 seconds
export const MAX_EXPIRATION_TIME = 672; // hours
export const PAGE_SIZE = 50;
export const ELEMENTS_CACHE_REQUEST_SIZE = 100;
export const DEFAULT_PLACEHOLDERS_COUNT = PAGE_SIZE;
export const ATTACHMENT_MAX_SIZE = 25000000; // bytes -> 25MB
export const LARGE_KEY_SIZE = 50 * 1024;
export const LOAD_RETRY_COUNT = 3;
export const LOAD_RETRY_DELAY = 3000; // in ms => 3s
export const PREVENT_CANCEL_SEND_INTERVAL = 30000; // Prevent form cancelling a message about to be sent 30s before

export const UNDO_SEND_DELAY = 5000;

export const SCHEDULED_MESSAGES_LIMIT = 100;
export const SCHEDULED_MAX_DATE_DAYS = 30;

export const MIN_DELAY_SENT_NOTIFICATION = 2500;

export const ELEMENT_TYPES = {
    MESSAGE: 'message',
    CONVERSATION: 'conversation',
};

export const LABEL_IDS_TO_HUMAN = {
    [MAILBOX_LABEL_IDS.INBOX]: 'inbox',
    [MAILBOX_LABEL_IDS.ALL_DRAFTS]: 'all-drafts',
    [MAILBOX_LABEL_IDS.ALL_SENT]: 'all-sent',
    [MAILBOX_LABEL_IDS.TRASH]: 'trash',
    [MAILBOX_LABEL_IDS.SPAM]: 'spam',
    [MAILBOX_LABEL_IDS.ALL_MAIL]: 'all-mail',
    [MAILBOX_LABEL_IDS.ARCHIVE]: 'archive',
    [MAILBOX_LABEL_IDS.SENT]: 'sent',
    [MAILBOX_LABEL_IDS.DRAFTS]: 'drafts',
    [MAILBOX_LABEL_IDS.STARRED]: 'starred',
    [MAILBOX_LABEL_IDS.OUTBOX]: 'outbox',
    [MAILBOX_LABEL_IDS.SCHEDULED]: 'scheduled',
};

export const HUMAN_TO_LABEL_IDS = Object.entries(LABEL_IDS_TO_HUMAN).reduce((acc, [key, value]) => {
    acc[value] = key;
    return acc;
}, Object.create(null));

export const getLabelIDsToI18N = () => ({
    [MAILBOX_LABEL_IDS.INBOX]: c('Link').t`Inbox`,
    [MAILBOX_LABEL_IDS.ALL_DRAFTS]: c('Link').t`Drafts`,
    [MAILBOX_LABEL_IDS.ALL_SENT]: c('Link').t`Sent`,
    [MAILBOX_LABEL_IDS.TRASH]: c('Link').t`Trash`,
    [MAILBOX_LABEL_IDS.SPAM]: c('Link').t`Spam`,
    [MAILBOX_LABEL_IDS.ALL_MAIL]: c('Link').t`All mail`,
    [MAILBOX_LABEL_IDS.ARCHIVE]: c('Link').t`Archive`,
    [MAILBOX_LABEL_IDS.SENT]: c('Link').t`Sent`,
    [MAILBOX_LABEL_IDS.DRAFTS]: c('Link').t`Drafts`,
    [MAILBOX_LABEL_IDS.STARRED]: c('Link').t`Starred`,
    [MAILBOX_LABEL_IDS.OUTBOX]: c('Link').t`Outbox`,
    [MAILBOX_LABEL_IDS.SCHEDULED]: c('Link').t`Scheduled`,
});

// List of location where messages are marked automatically as read after moving by the API
export const LABELS_AUTO_READ = [MAILBOX_LABEL_IDS.TRASH];

// List of location that cannot be change by user interaction
export const LABELS_UNMODIFIABLE_BY_USER = [
    MAILBOX_LABEL_IDS.ALL_MAIL,
    MAILBOX_LABEL_IDS.ALL_SENT,
    MAILBOX_LABEL_IDS.ALL_DRAFTS,
    MAILBOX_LABEL_IDS.OUTBOX,
    MAILBOX_LABEL_IDS.SCHEDULED,
];

export enum ENCRYPTED_STATUS {
    PGP_MIME = 8, // Used for attachment
}

export enum LINK_TYPES {
    WEB = 'web',
    EMAIL = 'email',
    PHONE = 'phone',
}

export enum MESSAGE_ACTIONS {
    NEW = -1,
    REPLY = 0,
    REPLY_ALL = 1,
    FORWARD = 2,
}

export const MAILTO_PROTOCOL_HANDLER_SEARCH_PARAM = `mailto`;
// Path: /inbox/#mailto=%s
export const MAILTO_PROTOCOL_HANDLER_PATH = `/${
    LABEL_IDS_TO_HUMAN[MAILBOX_LABEL_IDS.INBOX]
}/#${MAILTO_PROTOCOL_HANDLER_SEARCH_PARAM}=%s`;

export enum SAVE_DRAFT_ERROR_CODES {
    MESSAGE_ALREADY_SENT = 15034,
    DRAFT_DOES_NOT_EXIST = 15033,
}

export enum SEND_EMAIL_ERROR_CODES {
    MESSAGE_ALREADY_SENT = 2500,
}

export enum UPLOAD_ATTACHMENT_ERROR_CODES {
    MESSAGE_ALREADY_SENT = 11114,
}

export const MESSAGE_ALREADY_SENT_INTERNAL_ERROR = 'Message already sent';

export const DRAG_ELEMENT_KEY = 'drag-element';
export const DRAG_ELEMENT_ID_KEY = 'drag-element-id';
export const DRAG_ADDRESS_KEY = 'drag-address';
export const DRAG_ADDRESS_SIZE_KEY = 'drag-address-size';

export const MAX_ELEMENT_LIST_LOAD_RETRIES = 3;

export const OPENPGP_REFRESH_CUTOFF = 10;
export const ES_MAX_PARALLEL_MESSAGES = 150;
export const ES_MAX_CONCURRENT = 10;
export const ES_MAX_CACHE = 500000000; // 500 MB
export const ES_MAX_PAGES_PER_BATCH = 100;
export const ES_MAX_MESSAGES_PER_BATCH = 1000;
export const ES_MAX_INITIAL_CHARS = 20;
export const AesKeyGenParams: AesKeyGenParams = { name: 'AES-GCM', length: 128 };
export const KeyUsages: KeyUsage[] = ['encrypt', `decrypt`];
export const localisedForwardFlags = [
    'fw:',
    'fwd:',
    'vs:',
    'προώθ:',
    'tr :',
    'trs:',
    'áfram:',
    'welleh:',
    'en:',
    'enc:',
    'redir:',
    'vb:',
    '转发：',
    '轉寄：',
    '转发:',
    '轉寄:',
    'doorst:',
    'wg:',
    'πρθ:',
    'továbbítás:',
    'i:',
    'fs:',
    'rv:',
    'pd:',
    'i̇lt:',
];
export const defaultESStatus: ESStatus = {
    permanentResults: [],
    setElementsCache: () => {},
    labelID: '',
    lastEmail: undefined,
    previousNormSearchParams: undefined,
    cachedIndexKey: undefined,
    dbExists: false,
    isBuilding: false,
    isDBLimited: false,
    esEnabled: false,
    isRefreshing: false,
    isSearchPartial: false,
    isSearching: false,
    isCaching: false,
    isFirstSearch: true,
    dropdownOpened: false,
    temporaryToggleOff: false,
};
export const defaultESCache: ESCache = {
    esCache: [],
    cacheSize: 0,
    isCacheLimited: true,
    isCacheReady: false,
};
export const defaultESIndexingState: ESIndexingState = {
    esProgress: 0,
    estimatedMinutes: 0,
    startTime: 0,
    endTime: 0,
    oldestTime: 0,
    esPrevProgress: 0,
    totalIndexingMessages: 0,
    currentProgressValue: 0,
};

export const WELCOME_PANE_OPTIONS_URLS = {
    plansSelection: '/dashboard',
    protonShop: 'https://shop.protonmail.com',
    proton2FA: 'https://protonmail.com/support/knowledge-base/two-factor-authentication/',
    protonBusiness: 'https://protonmail.com/business/',
    calendar: 'https://calendar.protonmail.com',
    vpn: 'https://account.protonmail.com/u/6/vpn/vpn-apps',
};

export const restoringEncryptedMessagesURL =
    'https://protonmail.com/support/knowledge-base/restoring-encrypted-mailbox/';
export const reActivateKeySettingsURL = '/encryption-keys';

export const emailTrackerProtectionURL = 'https://protonmail.com/support/email-tracker-protection';

// Used for main action such as "label as", "move to" and "mark as read/unread"
export const SUCCESS_NOTIFICATION_EXPIRATION = 7500;

export const EO_REDIRECT_PATH = '/eo';
export const EO_MESSAGE_REDIRECT_PATH = `${EO_REDIRECT_PATH}/message`;
export const EO_REPLY_REDIRECT_PATH = `${EO_REDIRECT_PATH}/reply`;

export const EO_MAX_REPLIES_NUMBER = 5;

// Keys used for secure session storage
export const EO_TOKEN_KEY = 'proton:eo_token';
export const EO_DECRYPTED_TOKEN_KEY = 'proton:eo_decrypted_token';
export const EO_PASSWORD_KEY = 'proton:eo_password';
