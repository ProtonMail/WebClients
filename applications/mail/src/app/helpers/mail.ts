// TODO: drop pmcrypto import, see https://jira.protontech.ch/projects/CRYPTO/issues/CRYPTO-101
// eslint-disable-next-line no-restricted-imports
import { parseMail } from 'pmcrypto/lib/message/parseMail';

/**
 * Overwrites the headers in baseHeader by the headers defined in extraHeaders.
 */
export const combineHeaders = async (baseHeader: string, extraHeaders: string) => {
    const { headers: parsedExtras } = await parseMail(extraHeaders);
    const extraHeaderKeys = Object.keys(parsedExtras);
    return extraHeaderKeys.reduce((inputHeader, key) => {
        const capsKey = key.replace(/(^|-)./g, (letter) => letter.toUpperCase());
        const values = parsedExtras[key];
        const headerLine = values.map((val) => `${capsKey}: ${val}`).join('\r\n');
        const outputHeader = inputHeader.replace(new RegExp(`^${key}:.*(?=$[^ ])`, 'im'), headerLine);
        if (outputHeader.indexOf(capsKey) === -1) {
            return `${inputHeader + headerLine}\r\n`;
        }
        return outputHeader;
    }, baseHeader);
};

/**
 * Splits the mail into header and body.
 */
export const splitMail = (mail: string) => {
    // double new line separates message body and the headers
    const headersEnd = mail.search(/\r?\n\r?\n/);
    if (headersEnd === -1) {
        return { headers: mail, body: '' };
    }
    const headers = mail.substring(0, headersEnd);
    // remove leading newlines
    const body = mail.substring(headersEnd).trimStart();
    return { headers, body };
};
