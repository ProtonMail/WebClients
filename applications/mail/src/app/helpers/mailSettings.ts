import { Location } from 'history';

import { MailSettings } from '@proton/shared/lib/interfaces';
import { VIEW_LAYOUT, VIEW_MODE } from '@proton/shared/lib/mail/mailSettings';

import { isSearch } from './elements';
import { isAlwaysMessageLabels } from './labels';
import { extractSearchParameters } from './mailboxUrl';

export const isColumnMode = ({ ViewLayout = VIEW_LAYOUT.COLUMN }: Partial<MailSettings> = {}) =>
    ViewLayout === VIEW_LAYOUT.COLUMN;

export const isConversationMode = (
    labelID = '',
    { ViewMode = VIEW_MODE.GROUP }: Partial<MailSettings> = {},
    location?: Location
) => {
    const searchParams = location ? extractSearchParameters(location) : {};

    return !isAlwaysMessageLabels(labelID) && ViewMode === VIEW_MODE.GROUP && !isSearch(searchParams);
};
