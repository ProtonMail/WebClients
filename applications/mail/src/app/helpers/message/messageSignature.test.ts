import { MailSettings, UserSettings } from '@proton/shared/lib/interfaces';
import { PM_SIGNATURE as PM_SIGNATURE_ENUM } from '@proton/shared/lib/mail/mailSettings';
import { getProtonMailSignature } from '@proton/shared/lib/mail/signature';
import { message } from '@proton/shared/lib/sanitize';

import { MESSAGE_ACTIONS } from '../../constants';
import {
    CLASSNAME_SIGNATURE_CONTAINER,
    CLASSNAME_SIGNATURE_EMPTY,
    CLASSNAME_SIGNATURE_USER,
    insertSignature,
} from './messageSignature';

const content = '<p>test</p>';
const signature = `
<strong>>signature</strong>`;
const mailSettings = { PMSignature: PM_SIGNATURE_ENUM.DISABLED } as MailSettings;
const userSettings = {} as UserSettings;

const PM_SIGNATURE = getProtonMailSignature();

describe('signature', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('insertSignature', () => {
        describe('rules', () => {
            it('should remove line breaks', () => {
                const result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect(result).toContain('<br><strong>');
            });

            it('should try to clean the signature', () => {
                const result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect(result).toContain('&gt;');
            });

            it('should add empty line before the signature', () => {
                const result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect(result).toMatch(new RegExp(`<div><br></div>\\s*<div class="${CLASSNAME_SIGNATURE_CONTAINER}`));
            });

            it('should add different number of empty lines depending on the action', () => {
                let result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect((result.match(/<div><br><\/div>/g) || []).length).toBe(1);
                result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.REPLY,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect((result.match(/<div><br><\/div>/g) || []).length).toBe(2);
                result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.REPLY,
                    { ...mailSettings, PMSignature: PM_SIGNATURE_ENUM.ENABLED },
                    userSettings,
                    undefined,
                    false
                );
                expect((result.match(/<div><br><\/div>/g) || []).length).toBe(3);
                result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.REPLY,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect((result.match(/<div><br><\/div>/g) || []).length).toBe(3);
                result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.REPLY,
                    { ...mailSettings, PMSignature: PM_SIGNATURE_ENUM.ENABLED },
                    userSettings,
                    undefined,
                    false
                );
                expect((result.match(/<div><br><\/div>/g) || []).length).toBe(4);
            });

            it('should append PM signature depending mailsettings', () => {
                let result = insertSignature(content, '', MESSAGE_ACTIONS.NEW, mailSettings, {}, undefined, false);
                expect(result).not.toContain(PM_SIGNATURE);
                result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.NEW,
                    { ...mailSettings, PMSignature: PM_SIGNATURE_ENUM.ENABLED },
                    userSettings,
                    undefined,
                    false
                );
                const sanitizedPmSignature = message(PM_SIGNATURE);
                expect(result).toContain(sanitizedPmSignature);
                let messagePosition = result.indexOf(content);
                let signaturePosition = result.indexOf(sanitizedPmSignature);
                expect(messagePosition).toBeGreaterThan(signaturePosition);
                result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.NEW,
                    { ...mailSettings, PMSignature: PM_SIGNATURE_ENUM.ENABLED },
                    userSettings,
                    undefined,
                    true
                );
                messagePosition = result.indexOf(content);
                signaturePosition = result.indexOf(sanitizedPmSignature);
                expect(messagePosition).toBeLessThan(signaturePosition);
            });

            it('should append user signature if exists', () => {
                let result = insertSignature(
                    content,
                    '',
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect(result).toContain(`${CLASSNAME_SIGNATURE_USER} ${CLASSNAME_SIGNATURE_EMPTY}`);
                result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    false
                );
                expect(result).toContain('signature');
                let messagePosition = result.indexOf(content);
                let signaturePosition = result.indexOf(signature);
                expect(messagePosition).toBeGreaterThan(signaturePosition);
                result = insertSignature(
                    content,
                    signature,
                    MESSAGE_ACTIONS.NEW,
                    mailSettings,
                    userSettings,
                    undefined,
                    true
                );
                messagePosition = result.indexOf(content);
                signaturePosition = result.indexOf('signature');
                expect(messagePosition).toBeLessThan(signaturePosition);
            });
        });

        describe('snapshots', () => {
            const protonSignatures = [false, true];
            const userSignatures = [false, true];
            const actions = [
                MESSAGE_ACTIONS.NEW,
                MESSAGE_ACTIONS.REPLY,
                MESSAGE_ACTIONS.REPLY_ALL,
                MESSAGE_ACTIONS.FORWARD,
            ];
            const isAfters = [false, true];

            protonSignatures.forEach((protonSignature) => {
                userSignatures.forEach((userSignature) => {
                    actions.forEach((action) => {
                        isAfters.forEach((isAfter) => {
                            const label = `should match with protonSignature ${protonSignature}, userSignature ${userSignature}, action ${action}, isAfter ${isAfter}`;
                            it(label, () => {
                                const result = insertSignature(
                                    content,
                                    userSignature ? signature : '',
                                    action,
                                    {
                                        PMSignature: protonSignature
                                            ? PM_SIGNATURE_ENUM.ENABLED
                                            : PM_SIGNATURE_ENUM.DISABLED,
                                    } as MailSettings,
                                    userSettings,
                                    undefined,
                                    isAfter
                                );
                                expect(result).toMatchSnapshot();
                            });
                        });
                    });
                });
            });
        });
    });
});
