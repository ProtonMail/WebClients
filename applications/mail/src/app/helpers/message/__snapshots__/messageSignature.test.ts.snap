// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action -1, isAfter false 1`] = `
"<div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action -1, isAfter true 1`] = `
"<p>test</p><div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 0, isAfter false 1`] = `
"<div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 0, isAfter true 1`] = `
"<p>test</p><div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 1, isAfter false 1`] = `
"<div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 1, isAfter true 1`] = `
"<p>test</p><div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 2, isAfter false 1`] = `
"<div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature false, action 2, isAfter true 1`] = `
"<p>test</p><div><br></div>
<div class="protonmail_signature_block protonmail_signature_block-empty">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action -1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action -1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 0, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 0, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 2, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature false, userSignature true, action 2, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    
            <div class="protonmail_signature_block-proton protonmail_signature_block-empty">
        
            </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action -1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action -1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 0, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 0, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 2, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature false, action 2, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user protonmail_signature_block-empty">
        
            </div>
    
            <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action -1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action -1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 0, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 0, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 1, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 1, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 2, isAfter false 1`] = `
"<div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div><p>test</p>"
`;

exports[`signature insertSignature snapshots should match with protonSignature true, userSignature true, action 2, isAfter true 1`] = `
"<p>test</p><div><br></div><div><br></div>
<div class="protonmail_signature_block">
    <div class="protonmail_signature_block-user">
        <br><strong>&gt;signature</strong>
    </div>
    <div><br></div>
    <div class="protonmail_signature_block-proton">
        Sent with <a target="_blank" href="https://proton.me/">Proton Mail</a> secure email.
    </div>
</div>
<div><br></div>"
`;
